README.markdown
This is a wrapper of the DebiTech SOAP API. It's intended to be API compatible with the DebiTech Java client but also supports a more developer friendly syntax :).

Installing
----

    gem install debitech_soap

Usage
----
 
This is how you would have used the DebiTech Java API:

    include_class "com.verifyeasy.server.VEServer"
    veserver = VEServer.get_instance("https://secure.incab.se/verify/server/merchant_name")

This is how you use DebitechSoap:

    require 'debitech_soap'
    veserver = DebitechSoap::API.new(:merchant => "merchant_name", :username => "api_user_name", :password => "api_user_password")

Supported arguments
----

Java style (see DebitechSoap::API::PARAMS.keys in lib/debitech_soap.rb):

    veserver.refund(1234567, 23456, 100, "extra")

Hash:

    veserver.refund(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra")

Custom methods
----

**valid_credentials?**: Returns **true** if the credentials work (calls "checkSwedishPersNo").

Return data
----

- An object with methods for each attribute (See DebitechSoap::RETURN_DATA).
- Each attribute has serveral methods, for example "infoCode" can also be "getInfoCode" or "get_info_code".
- If the return value is a number it will be converted to an integer.

Gotchas
----

- We have only used the following methods in production: askIf3DSEnrolled, authorize3DS, authorize, subscribeAndSettle, checkSwedishPersNo, authorizeAndSettle.
- The other methods should work, but have not been tested.

API docs
----

Get DIBSServerManual.pdf from the DIBS Manager.

CI
----

- DebitechSoap is continuously tested for compability with Ruby 1.9.3 (preview 1), Ruby 1.8.7 and JRuby 1.6.2.


CHANGELOG
0.0.6

Using a local WSDL to avoid potential issues with downloading it each time the library is used.

0.0.5

Added timeout handling. Timeouts now return 403 so that you can retry.

0.0.4

Ruby 1.9 compability. Support for both the standard library soap implementation and soap4r.

0.0.3

Added support for Java style method names.

0.0.2

Removed activesupport dependency.

0.0.1

Initial release.


spec/debitech_soap_spec.rb
require File.expand_path(File.join(File.dirname(__FILE__), '../lib/debitech_soap'))

class MockSoapResult
  class MockReturn
    def method_missing(*opts)
      SOAP::Mapping::Object.new
    end
  end

  def return
    @return ||= MockReturn.new
  end
end

class MockSoapResultRuby19
  class MockReturn < MockSoapResult::MockReturn; end

  def m_return
    @return ||= MockReturn.new
  end
end

describe DebitechSoap::API, "valid_credentials?" do
  
  before do
    @client = mock(Object)
    SOAP::WSDLDriverFactory.stub!(:new).and_return(mock(Object, :create_rpc_driver => @client))
  end

  it "should call checkSwedishPersNo with the credentials and a valid swedish social security number" do
    @client.should_receive(:checkSwedishPersNo).with(:shopName => "merchant_name", :userName => "api_user_name",
                                                     :password => "api_user_password", :persNo => "555555-5555").
                                                     and_return(mock(Object, :return => "true"))

    api = DebitechSoap::API.new(:merchant => "merchant_name", :username => "api_user_name", :password => "api_user_password")
    api.valid_credentials?.should == true
  end

  it "should return false if the service returns false" do
    @client.stub!(:checkSwedishPersNo).and_return(mock(Object, :return => "false"))
    api = DebitechSoap::API.new(:merchant => "merchant_name", :username => "api_user_name", :password => "api_user_password")
    api.valid_credentials?.should == false
  end

  it "should work with Ruby 1.9 SOAP API" do
    @client.stub!(:checkSwedishPersNo).and_return(mock(Object, :m_return => "true"))
    api = DebitechSoap::API.new
    api.valid_credentials?.should == true
  end

end

describe DebitechSoap::API, "calling a method with java-style arguments" do

  before do
    @client = mock(Object)
    SOAP::WSDLDriverFactory.stub!(:new).and_return(mock(Object, :create_rpc_driver => @client))
  end
  
  it "should map the arguments to a hash and call the corresponding SOAP method" do
    api = DebitechSoap::API.new(:merchant => "merchant_name", :username => "api_user_name", :password => "api_user_password")
    @client.should_receive("refund").with(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra",
                                          :shopName => "merchant_name", :userName => "api_user_name", :password => "api_user_password").
                                          and_return(MockSoapResult.new)
    api.refund(1234567, 23456, 100, "extra")
  end

  it "should camel case method names" do
    api = DebitechSoap::API.new(:merchant => "merchant_name", :username => "api_user_name", :password => "api_user_password")
    @client.should_receive("authorize3DS").with(:verifyID => 1234567, :paRes => "RES", :extra => "extra",
                                          :shopName => "merchant_name", :userName => "api_user_name", :password => "api_user_password").
                                          and_return(MockSoapResult.new)
    api.authorize_3ds(1234567, "RES", "extra")
  end

  it "should not keep old attributes when making subsequent api calls" do
    api = DebitechSoap::API.new(:merchant => "merchant_name", :username => "api_user_name", :password => "api_user_password")
    @client.should_receive("refund").with(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra",
                                          :shopName => "merchant_name", :userName => "api_user_name", :password => "api_user_password").
                                          and_return(MockSoapResult.new)
    @client.should_receive("authorize3DS").with(:verifyID => 1234567, :paRes => "RES", :extra => "extra",
                                          :shopName => "merchant_name", :userName => "api_user_name", :password => "api_user_password").
                                          and_return(MockSoapResult.new)
                                          
    api.refund(1234567, 23456, 100, "extra")
    api.authorize3DS(1234567, "RES", "extra")
  end

  it "should create a return object" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResult.new
    mock_soap_result.return.stub!(:resultText).and_return("success")
    @client.stub!("refund").and_return(mock_soap_result)
    api.refund(1234567, 23456, 100, "extra").resultText.should == "success"
  end

  it "should return nil when there is no data" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResult.new
    @client.stub!("refund").and_return(mock_soap_result)
    api.refund(1234567, 23456, 100, "extra").resultCode.should be_nil
  end

  it "should be able to access the data using getCamelCase, get_underscore and underscore methods" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResult.new
    mock_soap_result.return.stub!(:resultText).and_return("success")
    @client.stub!("refund").and_return(mock_soap_result)
    result = api.refund(1234567, 23456, 100, "extra")
    result.getResultText.should == "success"
    result.get_result_text.should == "success"
    result.result_text.should == 'success'
  end

  it "should convert the result to an integer when its a number" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResult.new
    mock_soap_result.return.stub!(:resultCode).and_return("100")
    @client.stub!("refund").and_return(mock_soap_result)
    result = api.refund(1234567, 23456, 100, "extra")
    result.resultCode.should == 100
    result.getResultCode.should == 100
    result.get_result_code.should == 100
  end

  it "should convert the result to an integer when its zero" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResult.new
    mock_soap_result.return.stub!(:resultCode).and_return("0")
    @client.stub!("refund").and_return(mock_soap_result)
    result = api.refund(1234567, 23456, 100, "extra")
    result.resultCode.should == 0
  end

  it "should work with Ruby 1.9 SOAP API" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResultRuby19.new
    mock_soap_result.m_return.stub!(:resultCode).and_return("0")
    @client.stub!("refund").and_return(mock_soap_result)
    result = api.refund(1234567, 23456, 100, "extra")
    result.resultCode.should == 0
  end

end

describe DebitechSoap::API, "calling a method with hash-style arguments" do

  before do
    @client = mock(Object)
    SOAP::WSDLDriverFactory.stub!(:new).and_return(mock(Object, :create_rpc_driver => @client))
  end

  it "should call the corresponding soap method" do
    api = DebitechSoap::API.new(:merchant => "merchant_name", :username => "api_user_name", :password => "api_user_password")
    @client.should_receive("refund").with(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra",
                                          :shopName => "merchant_name", :userName => "api_user_name", :password => "api_user_password").
                                          and_return(MockSoapResult.new)
    api.refund(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra")
  end

  it "should return data" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResult.new
    mock_soap_result.return.stub!(:resultText).and_return("success")
    @client.stub!("refund").and_return(mock_soap_result)
    api.refund(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra").getResultText.should == "success"
  end

  it "should work with Ruby 1.9 SOAP API" do
    api = DebitechSoap::API.new
    mock_soap_result = MockSoapResultRuby19.new
    mock_soap_result.m_return.stub!(:resultText).and_return("success")
    @client.stub!("refund").and_return(mock_soap_result)
    result = api.refund(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra")
    result.getResultText.should == "success"
  end

end

describe DebitechSoap::API, "handling exceptions" do

  before do
    @client = mock(Object)
    SOAP::WSDLDriverFactory.stub!(:new).and_return(mock(Object, :create_rpc_driver => @client))
  end

  it "should catch Timeout::Error and return 403" do
    api = DebitechSoap::API.new
    @client.stub!("refund").and_raise(Timeout::Error)
    result = api.refund(:verifyID => 1234567, :transID => 23456, :amount => 100, :extra => "extra")
    result.getResultCode.should == 403
    result.getResultText.should == "SOAP Timeout"
  end
  
end


lib/debitech_soap/string_extensions.rb
module DebitechSoap
  module StringExtensions

    module Underscore
      def underscore
        word = dup
        word.gsub!(/::/, '/')
        word.gsub!(/([A-Z]+)([A-Z][a-z])/,'\1_\2')
        word.gsub!(/([a-z\d])([A-Z])/,'\1_\2')
        word.tr!("-", "_")
        word.downcase!
        word
      end
    end

    module CamelCase
      def camelcase(first_letter_in_uppercase = true)
        if first_letter_in_uppercase
          self.to_s.gsub(/\/(.?)/) { "::#{$1.upcase}" }.gsub(/(?:^|_)(.)/) { $1.upcase }
        else
          self.to_s[0].chr.downcase + self.camelcase[1..-1]
        end
      end
    end

  end
end

unless String.methods.include?('underscore')
  String.class_eval do
    include DebitechSoap::StringExtensions::Underscore
  end
end

unless String.methods.include?('camelcase')
  String.class_eval do
    include DebitechSoap::StringExtensions::CamelCase
  end
end


lib/debitech_soap.rb
require 'soap/wsdlDriver'
require 'ostruct'
require 'debitech_soap/string_extensions'

module DebitechSoap
  class API

    RETURN_DATA = %w{aCSUrl acquirerAddress acquirerAuthCode acquirerAuthResponseCode acquirerCity acquirerConsumerLimit acquirerErrorDescription acquirerFirstName acquirerLastName acquirerMerchantLimit acquirerZipCode amount errorMsg infoCode infoDescription pAReqMsg resultCode resultText verifyID}

    PARAMS = { %w(settle)                     => ["verifyID", "transID", "amount", "extra"],
               %w(subscribeAndSettle subscribe_and_settle) \
                                              => ["verifyID", "transID", "data", "ip", "extra"],
               %w(authorize)                  => ["billingFirstName", "billingLastName", "billingAddress", "billingCity",
                                                  "billingCountry", "cc", "expM", "expY", "eMail", "ip", "data", "currency", "transID", "extra"],
               %w(authorizeAndSettle3DS authorize_and_settle_3ds) \
                                              => ["verifyID", "paRes", "extra"],
               %w(refund)                     => ["verifyID", "transID", "amount", "extra"],
               %w(askIf3DSEnrolled ask_if_3ds_enrolled) \
                                              => ["billingFirstName", "billingLastName", "billingAddress", "billingCity",
                                                  "billingCountry", "cc", "expM", "expY", "eMail", "ip", "data", "currency", "transID",
                                                  "httpAcceptHeader", "httpUserAgentHeader", "method", "referenceNo", "extra"],
               %w(authReversal auth_reversal) => ["verifyID", "amount", "transID", "extra"],
               %w(authorize3DS authorize_3ds) => ["verifyID", "paRes", "extra"],
               %w(subscribe)                  => ["verifyID", "transID", "data", "ip", "extra"],
               %w(authorizeAndSettle authorize_and_settle) \
                                              => ["billingFirstName", "billingLastName", "billingAddress", "billingCity", "billingCountry",
                                                  "cc", "expM", "expY", "eMail", "ip", "data", "currency", "transID", "extra"] }

    def initialize(opts = {})
      @api_credentials = {}
      @api_credentials[:shopName] = opts[:merchant]
      @api_credentials[:userName] = opts[:username]
      @api_credentials[:password] = opts[:password]

      disable_stderr do
        @client = SOAP::WSDLDriverFactory.new(File.join(File.dirname(__FILE__), "service.wsdl")).create_rpc_driver
      end

      define_java_wrapper_methods!
    end

    def valid_credentials?
      disable_stderr do
        return_value(@client.checkSwedishPersNo(@api_credentials.merge({ :persNo => "555555-5555" }))) == "true"
      end
    end

  private

    # We use mumboe-soap4r for Ruby 1.9 compatibility, but if you use this library in 1.8 without using that lib
    # (e.g. without bundle exec) then you might get the standard library SOAP lib instead.
    # The standard library SOAP lib uses "return" here, and mumboe-soap4r uses "m_return".
    def return_value(object)
      object.respond_to?(:m_return) ? object.m_return : object.return
    end

    def define_java_wrapper_methods!
      PARAMS.keys.flatten.each { |method|
        (class << self; self; end).class_eval do                          # Doc:
          define_method(method) do |*args|                                # def refund(*args)
            attributes = @api_credentials.clone

            if args.first.is_a?(Hash) 
              attributes.merge!(args.first)
            else
              parameter_order = api_signature(method).last
              args.each_with_index { |argument, i|
                attributes[parameter_order[i].to_sym] = argument
              }
            end            
            begin
              client_result = return_value(@client.send(api_signature(method).first.first, attributes))
            rescue Timeout::Error
              client_result = OpenStruct.new(:resultCode => 403, :resultText => "SOAP Timeout")
              return return_data(client_result)
            end
            return_data(client_result)
          end
        end
      }
    end

    def return_data(results)
      hash = {}

      RETURN_DATA.each { |attribute|
        result = results.send(attribute)
        unless result.is_a?(SOAP::Mapping::Object)
          result = result.to_i if integer?(result)
          hash[attribute] = result
          hash["get_" + attribute.underscore] = result
          hash["get" + attribute.camelcase] = result
          hash[attribute.underscore] = result
        end
      }
      OpenStruct.new(hash)
    end

    def integer?(result)
      result.to_i != 0 || result == "0"
    end

    def disable_stderr
      begin
        $stderr = File.open('/dev/null', 'w')
        yield
      ensure
        $stderr = STDERR
      end
    end

    def api_signature(method)
      PARAMS.find {|key,value| key.include?(method.to_s) }
    end

  end
end
